pnet_datalink/src/linux.rs

```diff
@@ -119,23 +122,71 @@ pub fn channel(network_interface: &NetworkInterface,
 
     // Enable promiscuous capture
     if unsafe {
-        libc::setsockopt(socket,
-                         linux::SOL_PACKET,
-                         linux::PACKET_ADD_MEMBERSHIP,
-                         (&pmr as *const linux::packet_mreq) as *const libc::c_void,
-                         mem::size_of::<linux::packet_mreq>() as u32)
-    } == -1 {
+        libc::setsockopt(
+            socket,
+            linux::SOL_PACKET,
+            linux::PACKET_ADD_MEMBERSHIP,
+            (&pmr as *const linux::packet_mreq) as *const libc::c_void,
+            mem::size_of::<linux::packet_mreq>() as u32,
+        )
+    } == -1
+    {
         let err = io::Error::last_os_error();
         unsafe {
             pnet_sys::close(socket);
         }
         return Err(err);
     }
 
+    // Enable packet fanout
+    if let Some(fanout) = config.fanout {
+        use super::FanoutType;
+        let mut typ = match fanout.fanout_type {
+            FanoutType::HASH => linux::PACKET_FANOUT_HASH,
+            FanoutType::LB => linux::PACKET_FANOUT_LB,
+            FanoutType::CPU => linux::PACKET_FANOUT_CPU,
+            FanoutType::ROLLOVER => linux::PACKET_FANOUT_ROLLOVER,
+            FanoutType::RND => linux::PACKET_FANOUT_RND,
+            FanoutType::QM => linux::PACKET_FANOUT_QM,
+            FanoutType::CBPF => linux::PACKET_FANOUT_CBPF,
+            FanoutType::EBPF => linux::PACKET_FANOUT_EBPF,
+        } as u32;
+        // set defrag flag
+        if fanout.defrag {
+            typ = typ | linux::PACKET_FANOUT_FLAG_DEFRAG;
+        }
+        // set rollover flag
+        if fanout.rollover {
+            typ = typ | linux::PACKET_FANOUT_FLAG_ROLLOVER;
+        }
+        // set uniqueid flag -- probably not needed atm..
+        // PACKET_FANOUT_FLAG_UNIQUEID -- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=4a69a864209e9ab436d4a58e8028ac96cc873d15
```

Comment:
Any reason to not support this?

Reply:
There's no way to request socket's group_id, and uniqueid sets it to some unused value. We then have to reuse that value for opening other sockets. And we can't.

At the moment I decided against exporting get_group_id() API - mostly because I don't know how it could be exposed nicely.